src/dos/program_more_output.cpp

```diff
@@ -133,31 +158,56 @@ UserDecision MoreOutputBase::DisplaySingleStream()
 			code = '\n';
 		}
 		WriteOut("%c", code);
+		last_was_cr = (code == '\n') && !is_ansi;
 
-		// Detect 'new line' due to character passing the last column
+		// Detect redirected command output
+		const auto current_row    = GetCurrentRow();
 		const auto current_column = GetCurrentColumn();
+		if (last_was_cr) {
+			// New line should move the cursor to the first column
+			// and to the next row (unless we are already on the last
+			// line) - if not, the output must have been redirected
+			if (current_row == previous_row && current_row < max_lines) {
+				is_output_redirected = true;
+			} else if (current_column) {
+				is_output_redirected = true;
+			}
+		}
+		if (!is_ansi && (current_column == previous_column) &&
+		    code != code_del && (code >= ' ' || code < 0)) {
+			// Alphanumeric character outside of ANSI sequence
+			// always changes the current column - if not, the
+			// output must have been redirected
+			is_output_redirected = true;
+		}
+
+		// Detect 'new line' due to character passing the last column
+		bool line_overflow = false;
 		if (!current_column && previous_column && code != code_cr &&
 		    code != code_lf) {
 			// The cursor just moved to new line due to too small
-			// screen width. If this is followed by new line, ignore
-			// it, so that it is possible to i. e. nicely display up
-			// to 80-character lines on a standard 80 column screen
-			skip_next_cr = true;
-			skip_next_lf = true;
-			new_line     = true;
+			// screen width (line overflow). If this is followed by
+			// a new line, ignore it, so that it is possible to i. e.
+			// nicely display up to 80-character lines on a standard
+			// 80 column screen
+			skip_next_cr  = true;
+			skip_next_lf  = true;
+			new_line      = true;
+			line_overflow = true;
 		}
 		previous_column = current_column;
 
 		// Update new line counter, decide if pause needed
-		if (new_line && current_row) {
+		if (new_line && previous_row) {
 			++line_counter;
 		}
 		if (code != '\n') {
 			was_prompt_recently = false;
 		}
-		if (is_last) {
+		if (is_last && !line_overflow) {
 			// Skip further processing (including possible user prompt)
-			// if we know no data is left
+			// if we know no data is left and no line overflow produced
+			// an 'artificial new line'
```

Comment:
Seems fine at the surface level, but I find it hard to really understand how the various flags control the logic, whether the order of `if/else if` statement evaluations matter, etc. Given that you're checking the state of multiple flags in quite a few conditional statements, I'm suspecting what you really have here is a bunch of states that you're encoding with different flag combinations.

Wouldn't making the state explicit in the form of a state-machine make the code easier to undestand and maintain? As opposed to having implicit states in the form of various "undocumented" flag-combinations. Could you introduce a state enum, then convert this whole thing into an FSM?

Reply:
Definitely not the first time the CLI and shell parsing code has gotten quite tricky.

Fortunately this is readable and safe (unlike previously, where we had states plus lots of C char* pointer arithmetic), none the less, it's still juggling a lot of bools and branches.

@FeralChild64 - is it possible to feed content to `program_more_*` without standing up a whole dosbox instance? It might be a good candidate for unit tests.

The unit tests would let you hit all those state permutations (and variations on them too, illegal content, etc).

Reply:
@johnnovak: I have changed the code to use a state variable - but this was a huge and risky change, it will take quite some time to test it.

@kcgen: The MORE text engine tracks the cursor position after printing out characters - it would be hard to test it using unit tests, one would need to have a good simulation of the whole screen output, including ANSI handling.

Reply:
> I have changed the code to use a state variable - but this was a huge and risky change, it will take quite some time to test it.

Great, I think it reads a lot better, and it's much easier to reason about the transitions of a single state variable than  trying to visualise all possible permutations of 5+ boolean flags in your head. Even if it requires a bit more testing effort short term, the long-term maintenance aspects and the increased code clarity (which reduces the chance of weird bugs) makes it well worth it.

Thanks for doing this change @FeralChild64!

Reply:
> The MORE text engine tracks the cursor position after printing out characters - it would be hard to test it using unit tests, one would need to have a good simulation of the whole screen output, including ANSI handling.

Thanks for explaining that - and yes, that does sound excessively involved for testing (and at the same time, helps me understand this better too).